File: python/linear_regression.py
import numpy as np

def LinearRegression(X, Y):

	"""

	linear regression of given data(X, Y) 
	y = ax + b

	parameters
	-------------
	X  : independent variable data (array form)
	Y  : dependent variable data (array form)

	return
	-------------
	a  : the slope of regression line
	b  : intercept
	r2  : coefficient of determination

	"""

	n = len(X)
	sum_XY = np.dot(X,Y); sum_X = sum(X); sum_Y = sum(Y); sum_X2 = sum(X**2); sum_Y2 = sum(Y**2)

	a = (n*sum_XY - sum_X*sum_Y) / (n*sum_X2 - (sum_X)**2)
	b = (sum_Y - a*sum_X) / n

	r2 = (n*sum_XY - sum_X*sum_Y)**2 / ((n*sum_X2 - sum_X**2) * (n*sum_Y2 - sum_Y**2))

	return a, b, r2

File: matlab/linear regression.m
% Sample data
x = [1, 2, 3, 4, 5];
y = [2, 4, 5, 4, 5];

% Calculate the mean of x and y
mean_x = mean(x);
mean_y = mean(y);

% Calculate the slope (b) and y-intercept (a)
numerator = 0;
denominator = 0;
for i = 1:length(x)
    numerator = numerator + (x(i) - mean_x) * (y(i) - mean_y);
    denominator = denominator + (x(i) - mean_x) ^ 2;
end
b = numerator / denominator;
a = mean_y - b * mean_x;

% Print the equation of the line
fprintf('y = %.2fx + %.2f\n', b, a);